Updated: Redefine AppProtocol field description and add new standard values

LiorLieberman opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
See PR title 
#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Updated: Redefine AppProtocol field description and add new standard values

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
Enhancement Issue: <https://github.com/kubernetes/enhancements/pull/3727>

LiorLieberman committed:
redefine app protocol and add standard values

LiorLieberman commented:
/cc aojea 
/cc robscott

LiorLieberman commented:
/sig network

aojea commented on api/openapi-spec/v3/apis__discovery.k8s.io__v1_openapi.json:
```diff
@@ -140,7 +140,7 @@
         "properties": {
           "appProtocol": {
-            "description": "appProtocol represents the application protocol for this port. This field follows standard Kubernetes label syntax. Un-prefixed names are reserved for IANA standard service names (as per RFC-6335 and https://www.iana.org/assignments/service-names). Non-standard protocols should use prefixed names such as mycompany.com/my-custom-protocol.",
+            "description": "Used as a hint for implementations to configure the protocol used between the implementation and the application it exposes. This field follows standard Kubernetes label syntax. Valid values are either:\n\n* Un-prefixed protocol names - reserved for IANA standard service names (as per RFC-6335 and https://www.iana.org/assignments/service-names).\n\n* Kubernetes standar ...
```
is introducing `\n\n` characters , should it?

LiorLieberman replied:
I do have extra space lines between different bullets. so yes

aojea commented on pkg/apis/core/types.go:
```diff
@@ -4077,11 +4077,21 @@ type ServicePort struct {
-	// Non-standard protocols should use prefixed names such as
+	//
+	// * Kubernetes standard names:
+	//   * 'k8s.io/http2' - http2 over cleartext, aka 'h2c'. https://www.rfc-editor.org/rfc/rfc7540
```
this is super confusing, http2 uses TLS by default, cleartext is not commonly used,  did we agree on this name or is up to discussion?

LiorLieberman replied:
Is http2 uses TLS by default? Reading the [RFC](https://www.rfc-editor.org/rfc/rfc7540#section-3.1) my understanding is that it negotiates by default to choose h2/h2c.

should not we use https for http2 over tls? and let the client/app decide whether it should upgrade from http to http2 or not ? (https://www.rfc-editor.org/rfc/rfc7540#section-3.3) 

I think istio works as I explained right? @howardjohn

liggitt replied:
> shouldn't we use https for http2 over tls? and let the client/app decide whether it should upgrade from http to http2 or not ? (https://www.rfc-editor.org/rfc/rfc7540#section-3.3)

by the same token, should we use `http` for http2 without TLS and let the client negotiate an upgrade to h2c?

`k8s.io/http2` meaning http2 _without_ TLS is scary ... I could easily see someone using that not realizing they possibly hinted to another app to attempt non-TLS connections

LiorLieberman replied:
Thats a fair argument. I will change it to ‘h2c’ as it was originally discussed.

i did not find the comment chain where I suggested to change to http2. Maybe it was offline discussions

cici37 commented:
/remove-sig api-machinery

liggitt commented:
routing to @thockin for post-enhancement-freeze review since he was involved in the KEP

LiorLieberman committed:
change k8s.io/http2 to k8s.io/h2c

cici37 commented:
/remove-sig api-machinery

LiorLieberman commented:
/assign @thockin

robscott commented:
/assign

LiorLieberman commented:
@robscott any chance you had time to take a look?

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4996,11 +4996,21 @@ type EndpointPort struct {
 	Protocol Protocol `json:"protocol,omitempty" protobuf:"bytes,3,opt,name=protocol,casttype=Protocol"`
 
-	// The application protocol for this port.
+	// Used as a hint for implementations to
```
Is there a reason you're line-breaking at 40?  It looks like a haiku. :)

LiorLieberman replied:
lol :) 
probably not, i have opened it a while ago so do not remember why I did it.
Is there a format I should follow though?  80?

robscott replied:
Yep, 80 is what I've seen most commonly in k8s

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4996,11 +4996,21 @@ type EndpointPort struct {
 	// +optional
 	Protocol Protocol `json:"protocol,omitempty" protobuf:"bytes,3,opt,name=protocol,casttype=Protocol"`
 
-	// The application protocol for this port.
```
I think I'd keep this sentence.  e.g. 

"""
The application protocol for this port.  This is used as a hint for implementations to offer richer behavior for protocols that they understand.
"""

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4996,11 +4996,21 @@ type EndpointPort struct {
 	// RFC-6335 and https://www.iana.org/assignments/service-names).
-	// Non-standard protocols should use prefixed names such as
+	//
+	// * Kubernetes standard names:
```
I'd say "Kubernetes-defined prefixed names".

These should all be "kubernetes.io", though.  k8s.io was never supposed to be part of real APIs :)

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4996,11 +4996,21 @@ type EndpointPort struct {
+	// * Kubernetes standard names:
+	//   * 'k8s.io/h2c' - http2 over cleartext as described in https://www.rfc-editor.org/rfc/rfc7540
+	//   * 'k8s.io/grpc' - grpc traffic - see https://github.com/grpc/grpc/blob/v1.51.1/doc/PROTOCOL-HTTP2.md
+	//   * 'k8s.io/tcp' - plain tcp traffic
```
Why do we need this?  Isn't `protocol: TCP` plus the absence of `appProtocol` an indicator of "plain" TCP ?

LiorLieberman replied:
Apparently no. 
I think nowadays it is up to the implementation, For example, as far as i remember, in Istio, in case of`protocol: TCP` and  absence of `appProtocol`  they try to sniff and guess the protocol. However, with `appProtocol: TCP` they treat it as plain TCP.
@howardjohn is that right?

howardjohn replied:
Yep

thockin replied:
so we're asking people to specify "TCP" twice in order to defend against implementations?  That seems...unpleasant.

If we need to add a hint here, can we call it "raw" or something, which works for all port-protocols and tells implementations "don't try to be clever".

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4996,11 +4996,21 @@ type EndpointPort struct {
-	// Non-standard protocols should use prefixed names such as
+	//
+	// * Kubernetes standard names:
+	//   * 'k8s.io/h2c' - http2 over cleartext as described in https://www.rfc-editor.org/rfc/rfc7540
```
"HTTP/2" ?

Also - do we need HTTP/2 over TLS ?

aojea replied:
"HTTP/2" implies TLS by default, see section 3 of the rfc
```
[3.1](https://www.rfc-editor.org/rfc/rfc7540#section-3.1).  HTTP/2 Version Identification

   The protocol defined in this document has two identifiers.

   o  The string "h2" identifies the protocol where HTTP/2 uses
      Transport Layer Security (TLS) [[TLS12](https://www.rfc-editor.org/rfc/rfc7540#ref-TLS12)].  This identifier is used
      in the TLS application-layer protocol negotiation (ALPN) extension
      [[TLS-ALPN](https://www.rfc-editor.org/rfc/rfc7540#ref-TLS-ALPN)] field and in any place where HTTP/2 over TLS is
      identified.

      The "h2" string is serialized into an ALPN protocol identifier as
      the two-octet sequence: 0x68, 0x32.

   o  The string "h2c" identifies the protocol where HTTP/2 is run over
      cleartext TCP.  This identifier is used in the HTTP/1.1 Upgrade
      header field and in any place where HTTP/2 over TCP is identified.

      The "h2c" string is reserved from the ALPN identifier space but
      describes a protocol that does not use TLS.

   Negotiating "h2" or "h2c" implies the use of the transport, security,
   framing, and message semantics described in this document.
```

thockin replied:
Sorry, I meant "HTTP/2" in the explanation instead of "http2".

Other part stadns - If we are defining k8s "h2c", should we also define "h2" ?

LiorLieberman replied:
My thinking was that for "over TLS" we should encourage implementations to use `https` follow the ALPN negotiation. Thats what also reflected in the KEP. 
I do not think it is smart to have such thing "TLS only with h2", it is highly depended on the client and the server.
But I am not an expert in this field, maybe @howardjohn or @robscott  would be able to elaborate ?

thockin replied:
Fix the strings.  We can always add new definitions if needed, seeing as it is literally a comment :)

Code-freeze....

robscott replied:
I think @LiorLieberman's right here. "h2" would be accessed via ALPN, I think it would be difficult/impossible for many implementations to _only_ support "TLS only with h2".

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4996,11 +4996,21 @@ type EndpointPort struct {
+	//   * 'k8s.io/grpc' - grpc traffic - see https://github.com/grpc/grpc/blob/v1.51.1/doc/PROTOCOL-HTTP2.md
+	//   * 'k8s.io/tcp' - plain tcp traffic
+	//
+	// * Other protocols should use prefixed names such as
```
"user-defined prefixed names"

LiorLieberman replied:
I'd say it is more of "implementation-defined" than user defined. objections ?

thockin replied:
I'm fine with that

thockin commented on staging/src/k8s.io/api/discovery/v1/types.go:
```diff
@@ -184,11 +184,21 @@ type EndpointPort struct {
 	Port *int32 `json:"port,omitempty" protobuf:"bytes,3,opt,name=port"`
 
-	// appProtocol represents the application protocol for this port.
+	// Used as a hint for implementations to
```
Same comments as above

thockin reviewed (commented):
Sorry to nit-pick, but the docs are the whole feature :)

thockin commented:
ping!  There's about 24 hours until code freeze

LiorLieberman committed:
address feedback

robscott commented:
"Pod got deleted unexpectedly"

/retest

robscott reviewed (commented):
Thanks @LiorLieberman! Mostly LGTM, just a couple tiny nits

howardjohn commented on pkg/apis/core/types.go:
```diff
@@ -4078,10 +4078,19 @@ type ServicePort struct {
+	//
+	// * Kubernetes-defined prefixed names:
+	//   * 'kubernetes.io/h2c' - http2 over cleartext as described in https://www.rfc-editor.org/rfc/rfc7540
+	//   * 'kubernetes.io/grpc' - grpc traffic - see https://github.com/grpc/grpc/blob/v1.51.1/doc/PROTOCOL-HTTP2.md
```
gRPC over TLS or over plaintext?

robscott replied:
It may be sufficient to state "gRPC over HTTP/2" to align with the corresponding spec: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md. That spec does not seem to explicitly require TLS, just that "The HTTP2 specification mandates the use of TLS 1.2 or higher when TLS is used with HTTP2".

LiorLieberman replied:
or you mean "gRPC over HTTP/2" as not "grpc-web"?

howardjohn replied:
That doc implies its over HTTP2-over-TLS, but practically speaking gRPC over h2c is a real thing used in a variety of places, so a small comment would go a long ways I think.

thockin commented on pkg/apis/core/types.go:
```diff
@@ -4078,10 +4078,19 @@ type ServicePort struct {
+	// * Kubernetes-defined prefixed names:
+	//   * 'kubernetes.io/h2c' - http2 over cleartext as described in https://www.rfc-editor.org/rfc/rfc7540
+	//   * 'kubernetes.io/grpc' - grpc traffic - see https://github.com/grpc/grpc/blob/v1.51.1/doc/PROTOCOL-HTTP2.md
+	//   * 'kubernetes.io/tcp' - plain tcp traffic
```
Copying from lost comment:

so we're asking people to specify "TCP" twice in order to defend against implementations? That seems...unpleasant.

If we need to add a hint here, can we call it "raw" or something, which works for all port-protocols and tells implementations "don't try to be clever".

LiorLieberman replied:
Had an offline chat with @robscott . I removed it from here as it probably worth a little bit more discussion after we convert it to raw for all port protocols.

I will amend the KEP accordingly and add it as a Follow-up.

I know from [this](https://github.com/kubernetes/enhancements/pull/3727#discussion_r1068613722) comment that it is probably useful to have such "raw" value but I think we better iterate here than squeeze undesired change on last minute.

thockin commented on pkg/apis/core/types.go:
```diff
@@ -4078,10 +4078,19 @@ type ServicePort struct {
+	//   * 'kubernetes.io/grpc' - grpc traffic - see https://github.com/grpc/grpc/blob/v1.51.1/doc/PROTOCOL-HTTP2.md
+	//   * 'kubernetes.io/tcp' - plain tcp traffic
+	//
+	// * Other protocols should use prefixed names such as
```
still weird linewrap

LiorLieberman committed:
Update staging/src/k8s.io/api/discovery/v1/types.go

Co-authored-by: Rob Scott <rob.scott87@gmail.com>

LiorLieberman commented:
@thockin @robscott @howardjohn I do not want to do many commits, I need to run the `make update` each time.

I understand there is a code freeze so happy to finalize most of the feedback and then commit the changes.

Open things we have for now:
- plain TCP vs RAW 
-- I see the confusion, how about `k8s.io/raw-tcp` ? (I am also ok with only `k8s.io/raw` but wanted to be more explicit) 

- new comment from John - do we need to distinguish between gRPC over tls and over cleartext?
-- I have not good answer, how common is grpc to support this two values for now?

Things I will fix in the same commit:
- lines format
- "prefixed names" --> "implementation-defined prefixed names"
- http2, tcp, and grpc to capital letters

thockin commented:
> plain TCP vs RAW
> -- I see the confusion, how about k8s.io/raw-tcp ? (I am also ok with only k8s.io/raw but wanted to be more explicit)

I'm advocating for just "kubernetes.io/raw" - it's effectively a sub-protocol, and "TCP" has already been denoted in `protocol`.  I could also leave it out and debate the naming after this goes in.

thockin commented:
We can fix things like gRPC over TLS after code freeze - the main goal is to get this in.  Make the cosmetic changes and ping.

fedebongio commented:
/triage accepted

LiorLieberman committed:
remove kubernetes.io/tcp and change wording

LiorLieberman commented:
@thockin @robscott Ping.

I removed `kubernetes.io/tcp` for this iteration (more info [here](https://github.com/kubernetes/kubernetes/pull/115433#discussion_r1136267401)).

Changed to capital letters and wording where needed.

robscott commented:
Thanks @LiorLieberman!

/lgtm

thockin commented:
/approve

robscott commented:
Flake on "Kubectl client kubectl subresource flag GET on status subresource of built-in type (node) returns identical info as GET on the built-in type"

/retest
